utils/frame/try-runtime/cli/src/commands/fast_forward.rs

```diff
@@ -0,0 +1,292 @@
+
+/// Configurations of the [`Command::FastForward`].
+#[derive(Debug, Clone, clap::Parser)]
+pub struct FastForwardCmd {
+	/// How many blocks should be processed. If `None`, then blocks will be produced and processed
+	/// in a loop.
+	#[arg(long)]
+	n_blocks: Option<u64>,
+
+	/// The state type to use.
+	#[command(subcommand)]
+	state: State,
+
+	/// The ws uri from which to fetch the block.
+	///
+	/// If `state` is `Live`, this is ignored. Otherwise, it must not be empty.
+	// todo: make use of clap's arg groups
+	#[arg(long, value_parser = crate::parse::url)]
+	block_ws_uri: Option<String>,
+
+	/// Which try-state targets to execute when running this command.
+	///
+	/// Expected values:
+	/// - `all`
+	/// - `none`
+	/// - A comma separated list of pallets, as per pallet names in `construct_runtime!()` (e.g.
+	///   `Staking, System`).
+	/// - `rr-[x]` where `[x]` is a number. Then, the given number of pallets are checked in a
+	///   round-robin fashion.
+	#[arg(long, default_value = "all")]
+	try_state: frame_try_runtime::TryStateSelect,
+}
+
+impl FastForwardCmd {
+	fn block_ws_uri(&self) -> &str {
+		match self.state {
+			State::Live(LiveState { ref uri, .. }) => &uri,
+			_ => self
+				.block_ws_uri
+				.as_ref()
+				.expect("Either `--block-uri` must be provided, or state must be `live`"),
+		}
+	}
+}
+
+/// Something that can create inherent data providers and pre-runtime digest.
+///
+/// It is possible for the caller to provide custom arguments to the callee by setting the
+/// `ExtraArgs` generic parameter.
+///
+/// This module already provides some convenience implementation of this trait for closures. So, it
+/// should not be required to implement
+#[async_trait::async_trait]
+pub trait BlockBuildingInfoProvider<Block: BlockT, ExtraArgs = ()> {
+	type InherentDataProviders: InherentDataProvider;
+
+	async fn get_inherent_providers_and_pre_digest(
+		&self,
+		parent_hash: Block::Hash,
+		extra_args: ExtraArgs,
+	) -> Result<(Self::InherentDataProviders, Vec<DigestItem>)>;
+}
+
+#[async_trait::async_trait]
+impl<F, Block, IDP, ExtraArgs, Fut> BlockBuildingInfoProvider<Block, ExtraArgs> for F
+where
+	Block: BlockT,
+	F: Fn(Block::Hash, ExtraArgs) -> Fut + Sync + Send,
+	Fut: std::future::Future<Output = Result<(IDP, Vec<DigestItem>)>> + Send + 'static,
+	IDP: InherentDataProvider + 'static,
+	ExtraArgs: Send + 'static,
+{
+	type InherentDataProviders = IDP;
+
+	async fn get_inherent_providers_and_pre_digest(
+		&self,
+		parent: Block::Hash,
+		extra_args: ExtraArgs,
+	) -> Result<(Self::InherentDataProviders, Vec<DigestItem>)> {
+		(*self)(parent, extra_args).await
+	}
+}
+
+/// Read the block number corresponding to `hash` with an RPC call to `ws_uri`.
+async fn get_block_number<Block: BlockT>(
+	hash: Block::Hash,
+	ws_uri: &str,
+) -> Result<NumberFor<Block>>
+where
+	Block::Header: DeserializeOwned,
+{
+	let rpc = ws_client(ws_uri).await?;
+	Ok(ChainApi::<(), Block::Hash, Block::Header, ()>::header(&rpc, Some(hash))
```

Comment:
I strongly recall that there exists an RPC call that directly converts the block hash to number for you.

Reply:
I cannot find anything in https://github.com/paritytech/substrate/tree/master/client/rpc-api

Reply:
https://polkadot.js.org/docs/substrate/rpc/#getblockhashblocknumber-blocknumber-blockhash

Reply:
this call is for the opposite conversion, isn't it?

Reply:
sorry 🙈